File: frontend/src/services/userService.ts
import { gql } from "@apollo/client/core";
import { apolloClient } from "../apollo-client";

export interface User {
  id: string;
  username: string;
  email: string;
  role: "ADMIN" | "USER";
  createdAt: string;
}

export const userService = {
  getUser: async (id: string) => {
    const { data } = await apolloClient.query({
      query: gql`
        query GetUser($id: ID!) {
          user(id: $id) {
            id
            username
            email
            role
            createdAt
          }
        }
      `,
      variables: { id },
    });
    return data.user;
  },

  getUsers: async () => {
    const { data } = await apolloClient.query({
      query: gql`
        query GetUsers {
          users {
            id
            username
            email
            role
            createdAt
          }
        }
      `,
    });
    return data.users;
  },

  login: async (email: string, password: string) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation Login($email: String!, $password: String!) {
          login(email: $email, password: $password) {
            id
            username
            email
            role
          }
        }
      `,
      variables: { email, password },
    });
    return data.login;
  },

  logout: async () => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation Logout {
          logout
        }
      `,
    });
    return data.logout;
  },

  createUser: async (
    username: string,
    email: string,
    password: string,
    role: "ADMIN" | "USER"
  ) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation CreateUser(
          $username: String!
          $email: String!
          $password: String!
          $role: Role
        ) {
          createUser(
            username: $username
            email: $email
            password: $password
            role: $role
          ) {
            id
            username
            email
            role
            createdAt
          }
        }
      `,
      variables: { username, email, password, role },
    });
    return data.createUser;
  },
};

File: flutter/lib/data/services/user_service.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:graphql_flutter/graphql_flutter.dart';
import 'package:quiz_app/data/lib/models/user_model.dart'; // Zakładam, że masz model User
import 'auth_service.dart';

class UserService {
  final GraphQLClient _client = AuthService().client;
  final secureStorage = const FlutterSecureStorage();

  Future<Context?> _getHeadersContext() async {
    final String? token = await secureStorage.read(key: "jwt");

    if (token != null && token.isNotEmpty) {
      return Context.fromList([
        HttpLinkHeaders(headers: {'Authorization': 'Bearer $token'}),
      ]);
    }
    return null;
  }

  Future<UserModel> getUser({required String id}) async {
    final String query = """
      query GetUser(\$id: ID!) {
        user(id: \$id) {
          id
          username
          email
          role
          createdAt
        }
      }
    """;

    final QueryOptions options = QueryOptions(
      document: gql(query),
      variables: {'id': id},
      fetchPolicy: FetchPolicy.networkOnly,
      context: await _getHeadersContext(),
    );

    final QueryResult result = await _client.query(options);

    if (result.hasException) {
      print(result.exception.toString());
      throw Exception('Failed to load user');
    }

    final dynamic userJson = result.data?['user'];
    if (userJson == null) {
      throw Exception('User not found');
    }

    return UserModel.fromMap(userJson);
  }
}
